### PR TITLE
Pull request: Bug in tablefill for latex

### PR DESCRIPTION
**Pull request by lboxell**
@stanfordquan,

The tablefill function wasn't working as intended for latex. I made some changes to get it to work for my purposes. Feel free to take and integrate as you see fit (You'll want to edit the code if you use it; it was just a first pass to get things working for what I was doing). Let me know if you have any questions.

### CODE REVIEW

File: gslab_fill/tablefill.py [hunk 1]
```diff
@@ -77,39 +77,73 @@ def parse_data(data):
...
     
 
-def insert_tables(args,tables):
+def insert_tables(args, tables):
+    if re.search('\.tex', args['template']):
+        lyx_text = insert_tables_latex(args, tables)
+    else:
+        lyx_text = open(args['template'], 'rU').readlines()
+        for n in range( len(lyx_text) ):
+            if lyx_text[n].startswith('name "tab:') or re.search('label{tab:', lyx_text[n]):
```
Comment (yuchuan2016) [inline gslab_fill/tablefill.py]: Is the second condition for tex?

File: gslab_fill/tablefill.py [hunk 2]
```diff
@@ -77,39 +77,73 @@ def parse_data(data):
...
+                tag = re.sub("[\}\"\n]", "", lyx_text[n].split(':')[1]).lower()
```
Comment (yuchuan2016) [inline gslab_fill/tablefill.py]: Is this for tex? Is it equivalent to the previous line
```tag = lyx_text[n].replace('name "tab:','').rstrip('"\n').lower()```

File: gslab_fill/tablefill.py [hunk 3]
```diff
@@ -77,39 +77,73 @@ def parse_data(data):
...
+                            if re.match('---', tables[tag][entry_count]):
+                                rounded_entry = '---'
+                            else:
+                                rounded_entry = round_entry(entry_tag, tables[tag][entry_count])
+                                if re.match('^.*#\d+,#', lyx_text[i]):
+                                    rounded_entry = insert_commas(rounded_entry)
+                            lyx_text[i] = lyx_text[i].replace('#' + entry_tag + '#', rounded_entry)
+                            entry_count+=1
+                    
+                        elif lyx_text[i] == '</lyxtabular>\n' or re.search('end{tabular}', lyx_text[i]):
```
Comment (yuchuan2016) [inline gslab_fill/tablefill.py]: Is the second condition for tex?

File: gslab_fill/tablefill.py [hunk 4]
```diff
@@ -77,39 +77,73 @@ def parse_data(data):
     return tables    
     
 
-def insert_tables(args,tables):
+def insert_tables(args, tables):
+    if re.search('\.tex', args['template']):
+        lyx_text = insert_tables_latex(args, tables)
+    else:
```
Comment (yuchuan2016) [inline gslab_fill/tablefill.py]: I would prefer to splitting the function to `insert_tables_lyx` and `insert_tables_latex` and call the corresponding function based on `re.search('\.tex', args['template'])` or `re.search('\.lyx', args['template'])`

File: test.log [hunk 5]
```diff
@@ -21,21 +20,295 @@ gencat/tests/test_unzipFiles.py ......
...
+gslab_scons/tests/test_log.py ....F.F..
+gslab_scons/tests/test_misc.py .......F..........
 gslab_scons/tests/test_release_function.py ....
-gslab_scons/tests/test_release_tools.py ....
+gslab_scons/tests/test_release_tools.py F..F
 gslab_scons/tests/test_size_warning.py .....
 
-========================== 123 passed in 1.28 seconds ==========================
+=================================== FAILURES ===================================
+_________________ testTextfill.test_tags_incorrectly_specified _________________
```
Comment (yuchuan2016) [inline test.log]: Why does this fail?

**Comment by yuchuan2016** [comment]
@stanfordquan , I implemented my above comments that splitting the `insert_tables` to `insert_tables_lyx` and `insert_tables_tex`. The tests finished correctly. But I do not find the test we write for `gslab_fill` testing a LaTeX template input. Should writing the test for LaTeX tablefill a part of this pull request? @lboxell , please correct me if I miss something. Thanks!

File: gslab_fill/tablefill.py (hunk 1 above)
Comment (lboxell) [inline gslab_fill/tablefill.py]: Yes.

File: gslab_fill/tablefill.py (hunk 2 above)
Comment (lboxell) [inline gslab_fill/tablefill.py]: Yea, you can probably revert the changes to the base `insert_tables` function because I was trying to make the switches throughout, but then gave up and made a new function because there were more changes than anticipated.

File: test.log (hunk 5 above)
Comment (lboxell) [inline test.log]: I didn't spend any time on the unit tests when I was working on things.

**Comment by lboxell** [comment]
@yuchuan2016,

I left a couple comments on the questions you had. You can probably just export the .lyx table formatting from the original lyx-based unit tests into base .tex and otherwise use essentially the same tests. If you want me to test it on what I was working on before, just let me know.

**Comment by yuchuan2016** [comment]
@stanfordquan , I have added the tests for latex tablefill. Can you take a look before we merge? Thanks!

**Comment by stanfordquan** [comment]
I'm getting one failure, @yuchuan2016 . Any tip?

```python

=================================== FAILURES ===================================
_________________ testTextfill.test_tags_incorrectly_specified _________________

self = <gslab_fill.tests.test_textfill.testTextfill testMethod=test_tags_incorrectly_specified>

    def test_tags_incorrectly_specified(self):
        with nostderrout():
            textfill(input    = '../../gslab_fill/tests/input/tags_incorrectly_named.log',
                     template = '../../gslab_fill/tests/input/textfill_template.lyx',
                     output   = './build/textfill_template_filled.lyx')

        log_remove_string = '. insert_tag'

        with self.assertRaises(AssertionError):
            log = '../../gslab_fill/tests/input/tags_incorrectly_named.log'
>           self.check_log_in_LyX(log, log_remove_string, "textfill_")

../../gslab_fill/tests/test_textfill.py:91:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../gslab_fill/tests/test_textfill.py:57: in check_log_in_LyX
    self.assertEqual( len(text.results),2 )
E   AssertionError: 0 != 2
```

File: gslab_fill/tests/test_tablefill.py [hunk 6]
```diff
@@ -23,19 +23,43 @@ def setUp(self):
...
-                                output   = './build/tablefill_template_filled.lyx')
-        self.assertIn('filled successfully', message)                        
-        tag_data = open('../../gslab_fill/tests/input/tablefill_template.lyx', 'rU').readlines()
-        filled_data = open('./build/tablefill_template_filled.lyx', 'rU').readlines()
-        self.assertEqual(len(tag_data), len(filled_data))
-        for n in range(len(tag_data)):
-            self.tag_compare(tag_data[n], filled_data[n]) 
-    
-    def tag_compare(self, tag_line, filled_line):
+        for ext in ['lyx', 'tex']:
```
Comment (stanfordquan) [inline gslab_fill/tests/test_tablefill.py]: @yuchuan2016 , how do you feel about adding comments for these functions? I know we seem to not have comments for any of the tests (and this is from old code) but this is gonna get increasingly difficult to maintain, and since you know them pretty well right now this may be an opportune time to do so.

**Comment by yuchuan2016** [comment]
@stanfordquan , can you take a look to see if it works now? I have not touched `testTextfill`. I think the failure comes from an added line in `gitignore`. Thanks!

As to the comments, I agree comments would help a lot in understanding and maintaining these library. Honestly, I still don't quite understand the details of these tests. The tests I write are just copy-paste from existing tests with some modification. But I can definitely add some comments on the basic structure of these tests. Do you think we need to work on these now？

**Comment by stanfordquan** [comment]
@yuchuan2016 I'm with you on that. We can leave that for #112.

**Comment by stanfordquan** [comment]
Same `textfill` error. I vote to pull and ignore until #112.

**Comment by yuchuan2016** [comment]
@stanfordquan , quite strange... Agree.
